_posts/2018-05-25-ecs-calculator.md

```diff
@@ -0,0 +1,138 @@
+|m5.2xlarge| 8192| 31377|
+|m5.4xlarge| 16384| 62960|
+|m5.12xlarge| 49152| 189294|
+|m5.24xlarge| 98304| 378665|
+|----------------------|
+|c5.large| 2048| 3714|
+|c5.xlarge| 4096| 7634|
+|c5.2xlarge| 8192| 15473|
+|c5.4xlarge| 16384| 31152|
+|c5.9xlarge| 36864| 70351|
+|c5.18xlarge| 73728| 140780|
+|----------------------|
+|d2.xlarge| 4096| 30664|
+|d2.2xlarge| 8192| 61408|
+|d2.4xlarge| 16384| 122881|
+|d2.8xlarge| 36864| 245857|
+
+</div>
+
+❗️You can also find this information in yaml format [here](/assets/files/2018-05-25-ecs-calculator/ec2-instance-list.yml) 👈
+
+The solution is to make the optimal choice of the EC2 instance type with its budgeted CPU and Memory units to efficiently fit ECS containers in the instance based on the data from the table. Therefore, it is important to first have some insight into the limits of CPU and Memory reservation.
+
+The CPU units can be reserved with soft limit because containers can burst above their provision. One container can burst above its allocated units if no other containers are taking up resources. Besides, containers also share their unallocated CPU units with other containers on the instance with the same ratio as their allocated amount.
+
+It is also useful to understand how your container can effectively use the memory resources.The ECS memory resource model is more complex then the CPU resource model and allows you to configure both memory reservations and memory limits. You can see the whole concept below:
+
+![memory](/assets/posts/2018-05-25-ecs-calculator/memory_resources.png)
+
+Let's assume your container normally uses 128 MBs of memory, but occasionally bursts to 256 MBs of memory for short periods of time. As an option, you can set a Memory Reservation of 128 MBs, and a Memory Limit of 300 MBs. This configuration would allow the container to only reserve 128 MBs of memory from the resources on the container instance, but also allows the container to consume more memory resources when needed.
+
+We can consider another scenario where you configure only Hard Memory Limit with no Memory Reservation. In this case the ECS container instance will deduct configured memory limit from memory resources of the container instance. We think that it is better to use exactly this option because by specifying only a Hard Memory Limit for your tasks you avoid running out of memory (ECS stops placing tasks on the instance, and docker kills any containers that try to go over the hard limit 😿), in other words, Hard Memory Limit exists for the optimal fit of containers in the instance.
+
+## How the data can help you to optimally scale and fit ECS containers in an EC2 instance?
+
+Based on the soft/hard limit of CPU/Memory reservation, the approach of solving our current issue is taken according to the following steps.
+* Obtaining the number of the Memory units required for your largest container.
+* Optimally choosing the EC2 instance type that can budget those Memory units.
```

Comment:
Optimally choosing the EC2 instance type that can host your largest container based on the memory units it requires.